ministryofjustice/analytics-platform-go-unidler
# Unidling status updates

# What

Apologies for the massive diff. It's mostly the updated vendor packages.

* **Added Server Sent Events during unidling**
  This lets users see the progress of the app unidling in a dynamically updated webpage.

* **Refactor for testing**
  Testing this should be easier by making an `interface` for the Kubernetes API client and creating a mock implementation to use in the tests.

* **Refactor to recommended best practices**
  [Eg: Don't Panic!](https://github.com/golang/go/wiki/CodeReviewComments)

* **Switch to use `go mod` for dependencies**
  Packages are vendored, so it doesn't cause any problems. Best to keep up to date with the recommended way of doing things.

# How to review

1. checkout and use `make run` to run the unidler on `localhost:8080`
2. browse to http://localhost:8080/
3. see an "Unidling localhost:8080" page, with dynamically updated list of Server Sent Event messages - this should be one item "Failed unidling 'localhost:8080': can't find ingress for host 'localhost:8080'"
4. browse to http://localhost:8080/?host=IDLED_HOST where IDLED_HOST is the hostname of an idled app, eg: http://localhost:8080/?host=andy-test.apps.dev.mojanalytics.xyz
5. see an "Unidling IDLED_HOST" page, with SSE messages dynamically updating

I'm still not very experienced in Go, so some things might not be idiomatic - please comment!
